src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Object/ObjectWithParameterizedConstructorConverter.ReadJson.cs

```diff
@@ -0,0 +1,694 @@
+                    state.Current.JsonPropertyInfo = jsonPropertyInfo;
+                    useExtensionProperty = true;
+                }
+                else
+                {
+                    reader.Skip();
+                    return;
+                }
+            }
+            else
+            {
+                if (jsonPropertyInfo.JsonPropertyName == null && !options.PropertyNameCaseInsensitive)
+                {
+                    // Prevent future allocs by caching globally on the JsonPropertyInfo which is specific to a Type+PropertyName
+                    // so it will match the incoming payload except when case insensitivity is enabled (which is handled above).
+                    jsonPropertyInfo.JsonPropertyName = unescapedPropertyName.ToArray();
+                }
+
+                useExtensionProperty = false;
+            }
+
+            if (!jsonPropertyInfo.ShouldDeserialize)
+            {
+                reader.Skip();
+                return;
+            }
+
+            if (!useExtensionProperty)
+            {
+                if (state.Current.FoundProperties == null)
+                {
+                    state.Current.FoundProperties =
+                        ArrayPool<ValueTuple<
+                            JsonPropertyInfo,
+                            JsonReaderState,
+                            long,
+                            byte[]?>>.Shared.Rent(PropertyNameCountCacheThreshold);
+                }
+                else  if (state.Current.FoundPropertyCount == state.Current.FoundProperties!.Length)
+                {
+                    // Case where we can't fit all the JSON properties in the rented pool, we have to grow.
+                    var newCache = ArrayPool<ValueTuple<JsonPropertyInfo, JsonReaderState, long, byte[]?>>.Shared.Rent(state.Current.FoundProperties!.Length * 2);
+                    state.Current.FoundProperties!.CopyTo(newCache, 0);
+
+                    ArrayPool<ValueTuple<JsonPropertyInfo, JsonReaderState, long, byte[]?>>.Shared.Return(state.Current.FoundProperties!, clearArray: true);
+                    state.Current.FoundProperties = newCache!;
+                }
+
+                byte[]? propertyNameArray;
+
+                if (jsonPropertyInfo.JsonPropertyName == null)
+                {
+                    propertyNameArray = unescapedPropertyName.ToArray();
+                }
+                else
+                {
+                    propertyNameArray = null;
+                }
+
+                state.Current.FoundProperties![state.Current.FoundPropertyCount++] = (
+                    jsonPropertyInfo,
+                    reader.CurrentState,
+                    reader.BytesConsumed,
+                    propertyNameArray);
+
+                reader.Skip();
+            }
+            else
+            {
+                reader.Read();
+
+                if (state.Current.JsonClassInfo.DataExtensionIsObject)
```

Comment:
With this implementation, reading the extension data alongside the properties would mean caching the reader state for each extra property seen which increases the likelyhood of having to grow the `FoundProperties(Async)` cache (related to the conversation in https://github.com/dotnet/runtime/pull/33444#discussion_r391712806).

Also, it would mean more `if` checks when reading properties after the object is constructed (on the second pass), which would slow down perf when there is no extension data property.